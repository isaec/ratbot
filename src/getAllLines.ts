/**
 * matches "line 1234" but not "line 1234 and"
 */
const singleLine = /lines?\s(\d+)(?:\s(?!-|\d|and|&)|$)/g;

const singleHyphenatedLine = /lines?\s(\d+)-(?!\d)/g;

const lineRange = /lines?\s(\d+)-(\d+)/g;

const lineAndLine = /lines?\s(\d+)\s(?:and|&|\+)\s(\d+)/g;

const runRegexes = (
  string: string,
  results: Set<number>,
  ...regexes: RegExp[]
) => {
  for (const regex of regexes) {
    for (const match of string.matchAll(regex)) {
      if (typeof match[1] === "string") {
        results.add(parseInt(match[1]));
      }
    }
  }
};

export const getAllLines = (text: string): Set<number> => {
  const lines: Set<number> = new Set();
  // run the regex that return a single line match
  runRegexes(text, lines, singleLine, singleHyphenatedLine);
  for (const match of text.matchAll(lineRange)) {
    if (typeof match[1] === "string" && typeof match[2] === "string") {
      const start = parseInt(match[1]);
      const end = parseInt(match[2]);
      for (let i = start; i <= end; i++) lines.add(i);
    }
  }
  for (const match of text.matchAll(lineAndLine)) {
    if (typeof match[1] === "string" && typeof match[2] === "string") {
      lines.add(parseInt(match[1]));
      lines.add(parseInt(match[2]));
    }
  }

  return lines;
};

export const accumulate = <T>(iter: Iterator<T>): T[] => {
  const result: T[] = [];
  let next = iter.next();
  while (!next.done) {
    result.push(next.value);
    next = iter.next();
  }
  return result;
};

export const setToSortedArray = <T>(set: Set<T>): T[] =>
  accumulate(set.values()).sort();
