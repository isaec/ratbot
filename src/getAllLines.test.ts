import { assert, expect, test } from "vitest";
import { getAllLines, setToSortedArray } from "./getAllLines";

test.each([
  "Seeking approval for lines 86 and 87 (parts to fix swerve modules)",
  "requesting approval for line 88 for wheels for second intake shooter roller to swap if needed",
  "requesting approval for lines 89-91 for wheels for second intake shooter roller to swap if needed",
  "requesting mentor approval for line 93 (edited)",
  " Seeking approval for line 96 for 1/4-20 drill tap bits",
  "Seeking approval for line 99-bolt extractors for falcon stripped screws",
  " Requesting approval for line 102-104 need belts for 2nd intake and spares.",
  "seeking approval for lines 112-114",
  " Requesting approval for lines 117-118 for prototyping bolts and wing nuts \n Currently the bolts are button head and I'm not too sure if those are better than socket heads and the ones on the sheet are fully threaded",
  "seeking approval line 122 & 124",
  "seeking approval for lines 125-127, bumper fabric for onseason and shoulder bolt for swerve (edited) ",
  " seeking approval for lines 134 and 135. are the 3 1/2 needles the correct types we need?",
  "req approval for lines 129-130 (surgical tubing, diff ID). Same ones ordered as last year",
  "seeking approval for lines 16-18 and 22",
])(`getAllLines(%s)`, (text) => {
  expect(setToSortedArray(getAllLines(text))).toMatchSnapshot();
});
