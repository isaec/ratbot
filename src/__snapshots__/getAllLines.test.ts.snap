// Vitest Snapshot v1

exports[`getAllLines( Requesting approval for line 102-104 need belts for 2nd intake and spares.) 1`] = `
[
  102,
  103,
  104,
]
`;

exports[`getAllLines( Requesting approval for lines 117-118 for prototyping bolts and wing nuts 
 Currently the bolts are button head and I'm not too sure if those are better than socket heads and the ones on the sheet are fully threaded) 1`] = `
[
  117,
  118,
]
`;

exports[`getAllLines( Seeking approval for line 96 for 1/4-20 drill tap bits) 1`] = `
[
  96,
]
`;

exports[`getAllLines( seeking approval for lines 134 and 135. are the 3 1/2 needles the correct types we need?) 1`] = `
[
  134,
  135,
]
`;

exports[`getAllLines(Seeking approval for line 99-bolt extractors for falcon stripped screws) 1`] = `
[
  99,
]
`;

exports[`getAllLines(Seeking approval for lines 86 and 87 (parts to fix swerve modules)) 1`] = `
[
  86,
  87,
]
`;

exports[`getAllLines(req approval for lines 129-130 (surgical tubing, diff ID). Same ones ordered as last year) 1`] = `
[
  129,
  130,
]
`;

exports[`getAllLines(requesting approval for line 88 for wheels for second intake shooter roller to swap if needed) 1`] = `
[
  88,
]
`;

exports[`getAllLines(requesting approval for lines 89-91 for wheels for second intake shooter roller to swap if needed) 1`] = `
[
  89,
  90,
  91,
]
`;

exports[`getAllLines(requesting mentor approval for line 93 (edited)) 1`] = `
[
  93,
]
`;

exports[`getAllLines(seeking approval for lines 16-18 and 22) 1`] = `
[
  16,
  17,
  18,
  22,
]
`;

exports[`getAllLines(seeking approval for lines 112-114) 1`] = `
[
  112,
  113,
  114,
]
`;

exports[`getAllLines(seeking approval for lines 125-127, bumper fabric for onseason and shoulder bolt for swerve (edited) ) 1`] = `
[
  125,
  126,
  127,
]
`;

exports[`getAllLines(seeking approval line 122 & 124) 1`] = `
[
  122,
  124,
]
`;
